### PR TITLE
Improved Test Stability

### PR DESCRIPTION
**Pull request by russjones**
**Purpose**

Tests in `lib/srv/*` and `lib/utils/*` were quite unreliable. They usually worked on Linux, sometimes worked in a container, and never worked on macOS. This caused problems when merging a PR or while development.

This PR fixes this by moving those tests into the integration package and removing redundant tests.

**Implementation**

* Removed tests from `lib/srv/sshserver_test.go` related OpenSSH interoperability.
* Removed tests from `lib/utils/proxy/proxy_test.go` related to HTTP Connect Tunneling.
* Added `TestInteroperability` to the integrations package that tests OpenSSH interoperability.
* Added `TestTwoClustersProxy` to the integration package that tests reverse tunneling through a proxy.

### CODE REVIEW

**Comment by russjones** [comment]
Hrm, works on Linux and macOS, still fixing a few things when running tests in Docker.

**Comment by russjones** [comment]
@gravitational-jenkins test this please

File: integration/helpers.go [hunk 1]
```diff
@@ -558,6 +560,86 @@ func (i *TeleInstance) Stop(removeData bool) error {
...
+
+	// write 200 OK to the source, but don't close the connection
+	resp := &http.Response{
+		Status:     "OK",
+		StatusCode: 200,
+		Proto:      "HTTP/1.1",
+		ProtoMajor: 1,
+		ProtoMinor: 0,
+	}
+	resp.Write(sconn)
```
Comment (klizhentas) [inline integration/helpers.go]: check the return value

File: integration/helpers.go [hunk 2]
```diff
@@ -558,6 +560,86 @@ func (i *TeleInstance) Stop(removeData bool) error {
...
+	count int
+}
+
+// ServeHTTP only accepts the CONNECT verb and will tunnel your connection to
+// the specified host. Also tracks the number of connections that it proxies for
+// debugging purposes.
+func (p *proxyServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// validate http connect parameters
+	if r.Method != http.MethodConnect {
+		http.Error(w, fmt.Sprintf("%v not supported", r.Method), http.StatusInternalServerError)
```
Comment (klizhentas) [inline integration/helpers.go]: we have roundtrip.ReplyJSON that can save you time

File: integration/helpers.go [hunk 3]
```diff
@@ -558,6 +560,86 @@ func (i *TeleInstance) Stop(removeData bool) error {
...
+		return
+	}
+	if r.Host == "" {
+		http.Error(w, fmt.Sprintf("host not set"), http.StatusInternalServerError)
```
Comment (klizhentas) [inline integration/helpers.go]: trace.WriteError

File: integration/helpers.go [hunk 4]
```diff
@@ -558,6 +560,86 @@ func (i *TeleInstance) Stop(removeData bool) error {
...
+		return
+	}
+
+	// hijack request so we can get underlying connection
+	hj, ok := w.(http.Hijacker)
+	if !ok {
+		http.Error(w, "unable to hijack connection", http.StatusInternalServerError)
```
Comment (klizhentas) [inline integration/helpers.go]: trace.WriteError

File: integration/helpers.go [hunk 5]
```diff
@@ -558,6 +560,86 @@ func (i *TeleInstance) Stop(removeData bool) error {
...
+		io.Copy(sconn, dconn)
+		done <- true
+	}()
+	go func() {
+		io.Copy(dconn, sconn)
+		done <- true
+	}()
+
+	// wait until done
+	<-done
```
Comment (klizhentas) [inline integration/helpers.go]: missing timeouts, so tests will hang forever.

File: integration/helpers.go [hunk 6]
```diff
@@ -558,6 +560,86 @@ func (i *TeleInstance) Stop(removeData bool) error {
...
+
+	// success, we're proxying data now
+	p.Lock()
+	p.count = p.count + 1
+	p.Unlock()
+
+	// copy from src to dst and dst to src
+	done := make(chan bool)
+	go func() {
+		io.Copy(sconn, dconn)
```
Comment (klizhentas) [inline integration/helpers.go]: this could hang forever if the remote side crashes.

Comment (klizhentas) [inline integration/helpers.go]: overall it feels like reimplementation of vulcand/oxy proxy with middleware, may be just use it?

File: integration/integration_test.go [hunk 7]
```diff
@@ -275,6 +279,84 @@ func (s *IntSuite) TestAudit(c *check.C) {
...
+			"",
+			"not a tty",
+			false,
+		},
+	}
+
+	for i, tt := range tests {
+		// create new teleport server
+		t := s.newTeleport(c, nil, true)
+		defer t.Stop(true)
```
Comment (klizhentas) [inline integration/integration_test.go]: defer will be called on the function return not on the next iteration, so that's now what you expected probably - so scope it out to a separate function to guarantee better behavior.

File: integration/integration_test.go [hunk 8]
```diff
@@ -275,6 +279,84 @@ func (s *IntSuite) TestAudit(c *check.C) {
...
+		outbuf := &bytes.Buffer{}
+		cl.Stdin = inbuf
+		cl.Stdout = outbuf
+		cl.Stderr = outbuf
+
+		// run command and wait 100 ms for it complete
+		go func() {
+			// don't check for err, because sometimes this process should fail
+			// with an error and that's what the test is checking for.
+			cl.SSH(context.TODO(), []string{tt.inCommand}, false)
```
Comment (klizhentas) [inline integration/integration_test.go]: why not use channel to sync up here instead of just sleep

Comment (klizhentas) [inline integration/integration_test.go]: pass context to the function and use it to timeout in the main thread:

```bash
ctx := context.WithTimeout()
go func() {
    cl.SSH(ctx)
}()
select {
   case <- time.After(...):
 }
````

File: integration/integration_test.go [hunk 9]
```diff
@@ -382,6 +464,17 @@ func (s *IntSuite) TestInvalidLogins(c *check.C) {
 // Then it executes an SSH command on A by connecting directly
 // to A and by connecting to B via B<->A tunnel
 func (s *IntSuite) TestTwoClusters(c *check.C) {
+	// start the http proxy, we need to make sure this was not used
+	ps := &proxyServer{}
+	ts := httptest.NewServer(ps)
+	defer ts.Close()
+
+	// clear out any proxy environment variables
+	os.Setenv("http_proxy", "")
```
Comment (klizhentas) [inline integration/integration_test.go]: enough repetition to use loop

File: lib/srv/sshserver_test.go [hunk 10]
```diff
@@ -291,254 +290,6 @@ func (s *SrvSuite) TestAgentForward(c *C) {
 	c.Fatalf("expected socket to be closed, still could dial after 150 ms")
 }
 
-// TestExecCommandLocalPtyAndRemotePty tests executing a command where you have
```
Comment (klizhentas) [inline lib/srv/sshserver_test.go]: why did you remove all these tests?

Comment (russjones) [inline lib/srv/sshserver_test.go]: I moved the relevant tests into the integration package. I went a little overboard here originally in testing things that really were not all that relevant and/or covered elsewhere.

Comment (klizhentas) [inline lib/srv/sshserver_test.go]: ok, sounds good to me then.

File: integration/helpers.go [hunk 11]
```diff
@@ -571,30 +573,30 @@ type proxyServer struct {
 func (p *proxyServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	// validate http connect parameters
 	if r.Method != http.MethodConnect {
-		http.Error(w, fmt.Sprintf("%v not supported", r.Method), http.StatusInternalServerError)
+		trace.WriteError(w, fmt.Errorf("%v not supported", r.Method))
```
Comment (klizhentas) [inline integration/helpers.go]: `trace.WriteError(w, trace.BadParameter("%v is not supported", r.Method))`

Comment (klizhentas) [inline integration/helpers.go]: please don't use fmt.Errorf as it produces not standard HTTP response codes, while `trace.BadParameter` will produce 400 Bad Request and `trace.AccessDenied` will produce 403

File: integration/helpers.go [hunk 12]
```diff
@@ -618,23 +620,33 @@ func (p *proxyServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+	go replicate(dconn, sconn)
 
-	// wait until done
-	<-done
-	<-done
+	// wait until done, error, or 10 second
+	select {
+	case <-time.After(10 * time.Second):
+	case err := <-errc:
+		if err != nil {
```
Comment (klizhentas) [inline integration/helpers.go]: could be just `io.EOF` what is OK actually and should not result in errors

File: integration/helpers.go [hunk 13]
```diff
@@ -618,23 +620,33 @@ func (p *proxyServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+			trace.WriteError(w, err)
+			return
+		}
+	}
 
 	// close the connections
-	sconn.Close()
-	dconn.Close()
+	err = sconn.Close()
```
Comment (klizhentas) [inline integration/helpers.go]: why not defer

File: integration/helpers.go [hunk 14]
```diff
@@ -618,23 +620,33 @@ func (p *proxyServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		trace.WriteError(w, err)
```
Comment (klizhentas) [inline integration/helpers.go]: again, could be already closed or EOF

**Review by klizhentas — APPROVED** [review]
